From 5a44d962b00b0bd468dfa3bf1d77c51bae6b8137 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:29:40 -0400
Subject: [net] openvswitch: Correctly report flow used times for first 5 minutes after boot

Message-id: <87e812f60ff5d698f0191cedaa148509b11fd4c9.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94027
O-Subject: [RHEL7.1 net PATCH 07/50] openvswitch: Correctly report flow used times for first 5 minutes after boot.
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit f9b8c4c8baded129535d82d74df8e87a7a369f54
Author: Ben Pfaff <blp@nicira.com>
Date:   Thu Mar 20 10:45:21 2014 -0700

    openvswitch: Correctly report flow used times for first 5 minutes after boot.

    The kernel starts out its "jiffies" timer as 5 minutes below zero, as
    shown in include/linux/jiffies.h:

      /*
       * Have the 32 bit jiffies value wrap 5 minutes after boot
       * so jiffies wrap bugs show up earlier.
       */
      #define INITIAL_JIFFIES ((unsigned long)(unsigned int) (-300*HZ))

    The loop in ovs_flow_stats_get() starts out with 'used' set to 0, then
    takes any "later" time.  This means that for the first five minutes after
    boot, flows will always be reported as never used, since 0 is greater than
    any time already seen.

    Signed-off-by: Ben Pfaff <blp@nicira.com>
    Acked-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: Jesse Gross <jesse@nicira.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/flow.c b/net/openvswitch/flow.c
index a422471..2998989 100644
--- a/net/openvswitch/flow.c
+++ b/net/openvswitch/flow.c
@@ -92,7 +92,7 @@ static void stats_read(struct flow_stats *stats,
          unsigned long *used, __be16 *tcp_flags)
 {
  spin_lock(&stats->lock);
- if (time_after(stats->used, *used))
+ if (!*used || time_after(stats->used, *used))
   *used = stats->used;
  *tcp_flags |= stats->tcp_flags;
  ovs_stats->n_packets += stats->packet_count;
-- 
1.7.1